commit 1853eeb5e98b0ce26b9c336c17a05e2f614f8dfa
Author: Russell Brown <russell.brown@joyent.com>
Date:   2019-03-13T17:25:40+00:00 (7 months ago)
    
    MANTA-3519 Relocate picker unit test
    Reviewed by: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
    Approved by: Kelly McLaughlin <kelly.mclaughlin@joyent.com>

diff --git a/lib/picker.js b/lib/picker.js
index 9a83581..30db397 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -649,7 +649,10 @@ Picker.prototype.toString = function toString() {
     str += 'operatorDatacenters=' + this.operatorDatacenters.length + ', ';
     str += 'storageInterval=' + this.storageInterval + ', ';
     str += 'lag=' + this.lag + ', ';
-    str += 'moray=' + this.client.toString();
+    if (this.client) {
+        // i.e. NOT initialised in `standalone` mode
+        str += 'moray=' + this.client.toString();
+    }
     str += '>]';
 
     return (str);
@@ -667,133 +670,3 @@ module.exports = {
     sortAndStoreDcs: sortAndStoreDcs
 
 };
-
-
-
-///--- Tests
-
-var _cached_stub_data;
-function fetch_stub(opts, cb) {
-    cb = once(cb);
-
-    var fs = require('fs');
-    var path = require('path');
-
-    var fname = process.env.UNIT_TEST_STUB_NAME || 'picker.stub.json';
-    var file = path.join(__dirname, '..', 'test', fname);
-    var _opts = {
-        encoding: 'utf8'
-    };
-
-    if (_cached_stub_data) {
-        process.nextTick(function () {
-            cb(null, _cached_stub_data);
-        });
-    } else {
-        fs.readFile(file, _opts, function (err, data) {
-            if (err) {
-                cb(err);
-                return;
-            }
-
-            var values;
-            try {
-                values = JSON.parse(data).filter(function (obj) {
-                    return (obj.value.percentUsed < opts.utilization);
-                }).map(function (obj) {
-                    return (obj.value);
-                });
-            } catch (e) {
-                cb(e);
-                return;
-            }
-
-            _cached_stub_data = values;
-            cb(null, values);
-        });
-    }
-}
-
-
-
-function test(N) {
-    var picker = new Picker({
-        log: require('bunyan').createLogger({
-            level: process.env.LOG_LEVEL || 'info',
-            name: 'picker_test',
-            stream: process.stdout
-        }),
-        storageInterval: 10,
-        multiDC: true,
-        url: 'tcp://10.99.99.44:2020'
-    });
-
-    var min = Infinity;
-    var max = 0;
-    var total = 0;
-    var runs = 0;
-    var dcs = {};
-    var hosts = {};
-
-    function report() {
-        var keys = Object.keys(hosts);
-        keys.sort();
-        console.log('**** Host Selection ****');
-        keys.forEach(function (k) {
-            console.log(k + ' ' + hosts[k]);
-        });
-        console.log('\n**** Datacenter Selection ****');
-        keys = Object.keys(dcs);
-        keys.sort();
-        keys.forEach(function (k) {
-            console.log(k + ' ' + dcs[k]);
-        });
-        console.log('\n**** Timing ****');
-        console.log('avg: ' + total / N  + 'ms');
-        console.log('max: ' + max + 'ms');
-        console.log('min: ' + min + 'ms');
-
-        picker.close();
-    }
-
-    function select(t) {
-        var start = new Date().getTime();
-        picker.choose({}, function onChosen(err, sharks) {
-            assert.ifError(err);
-            var delta = new Date().getTime() - start;
-            if (delta > max)
-                max = delta;
-            if (delta < min)
-                min = delta;
-            total += delta;
-
-            Object.keys(sharks).forEach(function track(k) {
-                var dc_names = [];
-                sharks[k].forEach(function (s) {
-                    var id = s.manta_storage_id;
-                    if (!hosts[id])
-                        hosts[id] = 0;
-
-                    hosts[id]++;
-
-                    dc_names.push(s.datacenter);
-                });
-
-                var dc_key = dc_names.join(' ');
-                if (!dcs[dc_key])
-                    dcs[dc_key] = 0;
-                dcs[dc_key]++;
-            });
-
-            setImmediate(++runs < N ? select : report);
-        });
-    }
-
-    picker.on('connect', select);
-}
-
-
-if (process.env.UNIT_TEST) {
-    fetch = fetch_stub;
-    test(parseInt(process.argv.length > 2 ? process.argv[2] : 10000, 10));
-}
diff --git a/test/picker.records.json b/test/picker.records.json
new file mode 100644
index 0000000..5fef5a8
--- /dev/null
+++ b/test/picker.records.json
@@ -0,0 +1,68 @@
+{
+  "us-east-1": [
+    {
+      "manta_storage_id": "10.stor.us-east.joyent.us",
+      "availableMB": 71644665,
+      "datacenter": "us-east-1"
+    },
+    {
+      "manta_storage_id": "12.stor.us-east.joyent.us",
+      "availableMB": 72207835,
+      "datacenter": "us-east-1"
+    },
+    {
+      "manta_storage_id": "9.stor.us-east.joyent.us",
+      "availableMB": 71818391,
+      "datacenter": "us-east-1"
+    },
+    {
+      "manta_storage_id": "11.stor.us-east.joyent.us",
+      "availableMB": 71730379,
+      "datacenter": "us-east-1"
+    }
+  ],
+  "us-east-2": [
+    {
+      "manta_storage_id": "2.stor.us-east.joyent.us",
+      "availableMB": 67212403,
+      "datacenter": "us-east-2"
+    },
+    {
+      "manta_storage_id": "3.stor.us-east.joyent.us",
+      "availableMB": 66582795,
+      "datacenter": "us-east-2"
+    },
+    {
+      "manta_storage_id": "4.stor.us-east.joyent.us",
+      "availableMB": 67322598,
+      "datacenter": "us-east-2"
+    },
+    {
+      "manta_storage_id": "1.stor.us-east.joyent.us",
+      "availableMB": 67114405,
+      "datacenter": "us-east-2"
+    }
+  ],
+  "us-east-3": [
+    {
+      "manta_storage_id": "5.stor.us-east.joyent.us",
+      "availableMB": 67090948,
+      "datacenter": "us-east-3"
+    },
+    {
+      "manta_storage_id": "7.stor.us-east.joyent.us",
+      "availableMB": 67237238,
+      "datacenter": "us-east-3"
+    },
+    {
+      "manta_storage_id": "8.stor.us-east.joyent.us",
+      "availableMB": 66629966,
+      "datacenter": "us-east-3"
+    },
+    {
+      "manta_storage_id": "6.stor.us-east.joyent.us",
+      "availableMB": 66976466,
+      "datacenter": "us-east-3"
+    }
+  ]
+}
diff --git a/test/picker.stub.json b/test/picker.stub.json
deleted file mode 100644
index 7b89cf6..0000000
--- a/test/picker.stub.json
+++ /dev/null
@@ -1,435 +0,0 @@
-[
-  {
-    "bucket": "manta_storage",
-    "key": "bd7b792e-389c-4723-8bcf-635ac23eea00.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 71644665,
-      "percentUsed": 5,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 153595778089,
-        "bfree": 146728275397,
-        "bavail": 146728275397,
-        "files": 146729385848,
-        "ffree": 146728275397,
-        "favail": 146728275397,
-        "fsid": 23658767,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960993096,
-      "hostname": "bd7b792e-389c-4723-8bcf-635ac23eea00.stor.us-east.joyent.us",
-      "datacenter": "us-east-1",
-      "server_uuid": "00000000-0000-0000-0000-00259094c060",
-      "zone_uuid": "bd7b792e-389c-4723-8bcf-635ac23eea00",
-      "manta_compute_id": "26.cn.us-east.joyent.us",
-      "manta_storage_id": "10.stor.us-east.joyent.us"
-    },
-    "_id": 68,
-    "_etag": "4928000C",
-    "_mtime": 1381960997724,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "8386d8f5-d4ff-4b51-985a-061832b41179.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 67212403,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 153927359872,
-        "bfree": 137651003357,
-        "bavail": 137651003357,
-        "files": 137653149299,
-        "ffree": 137651003357,
-        "favail": 137651003357,
-        "fsid": 23658770,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960997057,
-      "hostname": "8386d8f5-d4ff-4b51-985a-061832b41179.stor.us-east.joyent.us",
-      "datacenter": "us-east-2",
-      "server_uuid": "00000000-0000-0000-0000-00259093ae14",
-      "zone_uuid": "8386d8f5-d4ff-4b51-985a-061832b41179",
-      "manta_compute_id": "12.cn.us-east.joyent.us",
-      "manta_storage_id": "2.stor.us-east.joyent.us"
-    },
-    "_id": 34,
-    "_etag": "D52B89CF",
-    "_mtime": 1381960997061,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "8e89d475-db39-49c8-b29a-7637e8bdec3d.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 72207835,
-      "percentUsed": 4,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154007791869,
-        "bfree": 147881647530,
-        "bavail": 147881647530,
-        "files": 147882734457,
-        "ffree": 147881647530,
-        "favail": 147881647530,
-        "fsid": 23658767,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960994140,
-      "hostname": "8e89d475-db39-49c8-b29a-7637e8bdec3d.stor.us-east.joyent.us",
-      "datacenter": "us-east-1",
-      "server_uuid": "70766742-a6cb-11e2-a84e-0025909584f0",
-      "zone_uuid": "8e89d475-db39-49c8-b29a-7637e8bdec3d",
-      "manta_compute_id": "295.cn.us-east.joyent.us",
-      "manta_storage_id": "12.stor.us-east.joyent.us"
-    },
-    "_id": 70,
-    "_etag": "1332E678",
-    "_mtime": 1381960994138,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "bcb0efe2-f4f1-4ed7-96fc-247656375fae.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 67090948,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154095214142,
-        "bfree": 137402262344,
-        "bavail": 137402262344,
-        "files": 137404295717,
-        "ffree": 137402262344,
-        "favail": 137402262344,
-        "fsid": 23658771,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960995773,
-      "hostname": "bcb0efe2-f4f1-4ed7-96fc-247656375fae.stor.us-east.joyent.us",
-      "datacenter": "us-east-3",
-      "server_uuid": "00000000-0000-0000-0000-00259094c090",
-      "zone_uuid": "bcb0efe2-f4f1-4ed7-96fc-247656375fae",
-      "manta_compute_id": "10.cn.us-east.joyent.us",
-      "manta_storage_id": "5.stor.us-east.joyent.us"
-    },
-    "_id": 28,
-    "_etag": "98735C3D",
-    "_mtime": 1381960995777,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "8b1bb7ac-036b-407c-a800-c1db372fa955.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 67237238,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154070745503,
-        "bfree": 137701863530,
-        "bavail": 137701863530,
-        "files": 137703905191,
-        "ffree": 137701863530,
-        "favail": 137701863530,
-        "fsid": 23658770,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960993882,
-      "hostname": "8b1bb7ac-036b-407c-a800-c1db372fa955.stor.us-east.joyent.us",
-      "datacenter": "us-east-3",
-      "server_uuid": "00000000-0000-0000-0000-00259094c044",
-      "zone_uuid": "8b1bb7ac-036b-407c-a800-c1db372fa955",
-      "manta_compute_id": "8.cn.us-east.joyent.us",
-      "manta_storage_id": "7.stor.us-east.joyent.us"
-    },
-    "_id": 30,
-    "_etag": "62A3FE8F",
-    "_mtime": 1381960993886,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "3d1e2fc2-5b9b-4c03-9560-821845899d38.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 66582795,
-      "percentUsed": 12,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 153979008428,
-        "bfree": 136361564624,
-        "bavail": 136361564624,
-        "files": 136363705581,
-        "ffree": 136361564624,
-        "favail": 136361564624,
-        "fsid": 23658770,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960986423,
-      "hostname": "3d1e2fc2-5b9b-4c03-9560-821845899d38.stor.us-east.joyent.us",
-      "datacenter": "us-east-2",
-      "server_uuid": "00000000-0000-0000-0000-00259094c038",
-      "zone_uuid": "3d1e2fc2-5b9b-4c03-9560-821845899d38",
-      "manta_compute_id": "19.cn.us-east.joyent.us",
-      "manta_storage_id": "3.stor.us-east.joyent.us"
-    },
-    "_id": 33,
-    "_etag": "0E1709EF",
-    "_mtime": 1381960995001,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "2a7157a0-e53f-4863-964e-fe6078305503.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 71818391,
-      "percentUsed": 5,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 153992244055,
-        "bfree": 147084065158,
-        "bavail": 147084065158,
-        "files": 147085185976,
-        "ffree": 147084065158,
-        "favail": 147084065158,
-        "fsid": 23658767,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960996797,
-      "hostname": "2a7157a0-e53f-4863-964e-fe6078305503.stor.us-east.joyent.us",
-      "datacenter": "us-east-1",
-      "server_uuid": "00000000-0000-0000-0000-00259094bf40",
-      "zone_uuid": "2a7157a0-e53f-4863-964e-fe6078305503",
-      "manta_compute_id": "25.cn.us-east.joyent.us",
-      "manta_storage_id": "9.stor.us-east.joyent.us"
-    },
-    "_id": 67,
-    "_etag": "45CA0127",
-    "_mtime": 1381960996802,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "d4ecaf94-ac9e-48d7-b9c6-0fc9257a818e.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 66629966,
-      "percentUsed": 12,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154012096503,
-        "bfree": 136458170849,
-        "bavail": 136458170849,
-        "files": 136460215741,
-        "ffree": 136458170849,
-        "favail": 136458170849,
-        "fsid": 23658769,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960995956,
-      "hostname": "d4ecaf94-ac9e-48d7-b9c6-0fc9257a818e.stor.us-east.joyent.us",
-      "datacenter": "us-east-3",
-      "server_uuid": "00000000-0000-0000-0000-00259094c058",
-      "zone_uuid": "d4ecaf94-ac9e-48d7-b9c6-0fc9257a818e",
-      "manta_compute_id": "9.cn.us-east.joyent.us",
-      "manta_storage_id": "8.stor.us-east.joyent.us"
-    },
-    "_id": 31,
-    "_etag": "713E9B71",
-    "_mtime": 1381960995960,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "5ecf3eae-317b-4be1-b081-ce88d73ee35f.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 66976466,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154021747843,
-        "bfree": 137167803541,
-        "bavail": 137167803541,
-        "files": 137169846238,
-        "ffree": 137167803541,
-        "favail": 137167803541,
-        "fsid": 23658770,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960996448,
-      "hostname": "5ecf3eae-317b-4be1-b081-ce88d73ee35f.stor.us-east.joyent.us",
-      "datacenter": "us-east-3",
-      "server_uuid": "00000000-0000-0000-0000-002590935960",
-      "zone_uuid": "5ecf3eae-317b-4be1-b081-ce88d73ee35f",
-      "manta_compute_id": "1.cn.us-east.joyent.us",
-      "manta_storage_id": "6.stor.us-east.joyent.us"
-    },
-    "_id": 29,
-    "_etag": "D3E610B7",
-    "_mtime": 1381960996452,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "31850a62-42b2-4439-81d6-b24b8c04c692.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 67322598,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154066528913,
-        "bfree": 137876681167,
-        "bavail": 137876681167,
-        "files": 137878848117,
-        "ffree": 137876681167,
-        "favail": 137876681167,
-        "fsid": 23658771,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960995874,
-      "hostname": "31850a62-42b2-4439-81d6-b24b8c04c692.stor.us-east.joyent.us",
-      "datacenter": "us-east-2",
-      "server_uuid": "00000000-0000-0000-0000-002590913d48",
-      "zone_uuid": "31850a62-42b2-4439-81d6-b24b8c04c692",
-      "manta_compute_id": "11.cn.us-east.joyent.us",
-      "manta_storage_id": "4.stor.us-east.joyent.us"
-    },
-    "_id": 32,
-    "_etag": "3E6EB0F0",
-    "_mtime": 1381960995884,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "544787c3-330c-40e7-bc47-30065fa0e650.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 71730379,
-      "percentUsed": 5,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 153976371801,
-        "bfree": 146903817371,
-        "bavail": 146903817371,
-        "files": 146904922228,
-        "ffree": 146903817371,
-        "favail": 146903817371,
-        "fsid": 23658768,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960990130,
-      "hostname": "544787c3-330c-40e7-bc47-30065fa0e650.stor.us-east.joyent.us",
-      "datacenter": "us-east-1",
-      "server_uuid": "6ea1aa0a-a6f1-11e2-8f76-00259094386c",
-      "zone_uuid": "544787c3-330c-40e7-bc47-30065fa0e650",
-      "manta_compute_id": "293.cn.us-east.joyent.us",
-      "manta_storage_id": "11.stor.us-east.joyent.us"
-    },
-    "_id": 69,
-    "_etag": "D5698A10",
-    "_mtime": 1381960998474,
-    "_txn_snap": null,
-    "_count": 12
-  },
-  {
-    "bucket": "manta_storage",
-    "key": "00c3e6bd-b8bc-4300-919b-bf36d4cd8920.stor.us-east.joyent.us",
-    "value": {
-      "availableMB": 67114405,
-      "percentUsed": 11,
-      "filesystem": "/manta",
-      "statvfs": {
-        "bsize": 131072,
-        "frsize": 512,
-        "blocks": 154070493707,
-        "bfree": 137450303354,
-        "bavail": 137450303354,
-        "files": 137452463272,
-        "ffree": 137450303354,
-        "favail": 137450303354,
-        "fsid": 23658777,
-        "basetype": "zfs",
-        "flag": 4,
-        "namemax": 255,
-        "fstr": ""
-      },
-      "timestamp": 1381960998349,
-      "hostname": "00c3e6bd-b8bc-4300-919b-bf36d4cd8920.stor.us-east.joyent.us",
-      "datacenter": "us-east-2",
-      "server_uuid": "00000000-0000-0000-0000-00259094c054",
-      "zone_uuid": "00c3e6bd-b8bc-4300-919b-bf36d4cd8920",
-      "manta_compute_id": "20.cn.us-east.joyent.us",
-      "manta_storage_id": "1.stor.us-east.joyent.us"
-    },
-    "_id": 27,
-    "_etag": "3CA43E51",
-    "_mtime": 1381960998353,
-    "_txn_snap": null,
-    "_count": 12
-  }
-]
-
diff --git a/test/picker.test.js b/test/picker.test.js
new file mode 100644
index 0000000..af13d43
--- /dev/null
+++ b/test/picker.test.js
@@ -0,0 +1,220 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2019, Joyent, Inc.
+ */
+
+var bunyan = require('bunyan');
+var mod_picker = require('../lib/picker.js');
+var fs = require('fs');
+var path = require('path');
+var assert = require('assert-plus');
+
+///--- Constants
+
+const THREE_SIGMA = 0.997;
+var DEF_MAX_STREAMING_SIZE_MB = 5120;
+var DEF_MAX_PERCENT_UTIL = 90;
+var DEF_MAX_OPERATOR_PERCENT_UTIL = 92;
+
+
+///--- Tests
+
+/**
+ * Sum values in array
+ *
+ * @param {!number[]} values
+ */
+function sum(values) {
+    return values.reduce(function (acc, v) {
+        return (acc += v);
+    }, 0);
+}
+
+/**
+ * Calculate the arithmetic mean of an array of values
+ *
+ * @param {!number[]} values
+ * @returns {number} the sum of values
+ */
+function mean(values) {
+    return (sum(values) / values.length);
+}
+
+/**
+ * Calculate the standard deviation of an array of values
+ *
+ * @param {!number[]} values
+ * @param {number} [av] - mean, if already calculated
+ * @returns {number} the standard deviation of values
+ */
+function sd(values, av) {
+    if (av === undefined) {
+        av = mean(values);
+    }
+
+    var sqd = values.map(function (v) {
+        var res = v - av;
+        return (res * res);
+    });
+
+    var meansqd = mean(sqd);
+    return (Math.sqrt(meansqd));
+}
+
+/**
+ * Calculate zscore of each element in array of values
+ *
+ * @param {!number[]} values
+ * @param {number} [av] - mean, if already calculated
+ * @param {number} [stdd] - standard deviation, if already calculated
+ * @returns {number[]} an array of zscores for values
+ */
+function zscores(values, av, stdd) {
+    if (av === undefined) {
+        av = mean(values);
+    }
+    if (stdd === undefined) {
+        stdd = sd(values, av);
+    }
+    return (values.map(function (v) {
+        return ((v - av) / stdd);
+    }));
+}
+
+/**
+ * Rough test to see if values follow a normal distribution
+ *
+ * NOTE: this is non-deterministic, we could do with some Erlang quickcheck
+ * ?SOMETIMES magic here (see http://quviq.com/documentation/eqc/ macro
+ * ?SOMETIMES for details.)
+ *
+ * @param {object} kvs - an object of {string} -> {number} mappings
+ * @returns {boolean} true if the numbers are most likely normally
+ * distributed
+ */
+function probablyNormal(kvs) {
+    var values = Object.keys(kvs).map(function (k) {
+        return (kvs[k]);
+    });
+    var zs = zscores(values);
+
+    // outliers are greater than 3sd from the mean
+    var outliers = zs.filter(function (z) {
+        return (Math.abs(z) > 3);
+    });
+
+    //NOTE: in this case, probably outliers.length === 0 would be
+    //enough
+    return ((outliers.length / values.length) < (1 - THREE_SIGMA));
+}
+
+exports.pickerTest = function (t) {
+    // report vars (see report below)
+    var min = Infinity;
+    var max = 0;
+    var total = 0;
+    var runs = 0;
+    var dcs = {};
+    var hosts = {};
+
+    // Test iterations. Pulled from old lib/picker.js test, see
+    // blame/history on that file
+    var N = 10000;
+
+    // read sub/mock/fake moray results
+    var fname = 'picker.records.json';
+    var file = path.join(__dirname, '..', 'test', fname);
+    var morayValuesJSON = fs.readFileSync(file, 'utf8');
+    var morayValues = JSON.parse(morayValuesJSON);
+
+    // make a picker
+    var picker = mod_picker.createClient({
+        log: require('bunyan').createLogger({
+            level: process.env.LOG_LEVEL || 'info',
+            name: 'picker_test',
+            stream: process.stdout
+        }),
+        defaultMaxStreamingSizeMB: DEF_MAX_OPERATOR_PERCENT_UTIL,
+        maxUtilizationPct: DEF_MAX_PERCENT_UTIL,
+        maxOperatorUtilizationPct: DEF_MAX_OPERATOR_PERCENT_UTIL,
+        multiDC: true,
+        // i.e. don't connect to moray
+        standalone: true
+    });
+
+    // set up the picker with the fake moray data
+    mod_picker.sortAndStoreDcs.call(picker, morayValues, morayValues);
+
+    // do it
+    select(t);
+
+    // NOTE: only reports on failure, to aid debugging
+    function report() {
+        var hostsNormal = probablyNormal(hosts);
+        var dcsNormal = probablyNormal(dcs);
+
+        if (!(hostsNormal && dcsNormal)) {
+            var keys = Object.keys(hosts);
+            keys.sort();
+            console.log('**** Host Selection ****');
+            keys.forEach(function (k) {
+                console.log(k + ' ' + hosts[k]);
+            });
+            console.log('\n**** Datacenter Selection ****');
+            keys = Object.keys(dcs);
+            keys.sort();
+            keys.forEach(function (k) {
+                console.log(k + ' ' + dcs[k]);
+            });
+            console.log('\n**** Timing ****');
+            console.log('avg: ' + total / N  + 'ms');
+            console.log('max: ' + max + 'ms');
+            console.log('min: ' + min + 'ms');
+        }
+
+        picker.close();
+        t.ok(hostsNormal, 'Hosts selection appears normally distributed');
+        t.ok(dcsNormal, 'DC selection appears normally distributed');
+        t.done();
+    }
+
+    // the actual test code, yes, it runs iterations
+    function select() {
+        var start = new Date().getTime();
+        picker.choose({}, function onChosen(err, sharks) {
+            assert.ifError(err);
+            var delta = new Date().getTime() - start;
+            if (delta > max)
+                max = delta;
+            if (delta < min)
+                min = delta;
+            total += delta;
+
+            Object.keys(sharks).forEach(function track(k) {
+                var dc_names = [];
+                sharks[k].forEach(function (s) {
+                    var id = s.manta_storage_id;
+                    if (!hosts[id])
+                        hosts[id] = 0;
+
+                    hosts[id]++;
+
+                    dc_names.push(s.datacenter);
+                });
+
+                var dc_key = dc_names.join(' ');
+                if (!dcs[dc_key])
+                    dcs[dc_key] = 0;
+                dcs[dc_key]++;
+            });
+
+            setImmediate((++runs < N ? select : report), t);
+        });
+    }
+
+};
